File: WiFiHeatMap.Tests/SignalServiceTests.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using NSubstitute;
using WiFiHeatMap.Server;
using Xunit;

namespace WiFiHeatMap.Tests
{
    public class SignalServiceTests
    {
        [Fact]
        public async Task SetsStatusOnException()
        {
            //arrange
            var reader = Substitute.For<ISignalReader<string>>();
            reader.When(r => r.Read()).Throw(new Exception("unit test exception"));
            var parser = Substitute.For<ISignalParser<string>>();
            var hub = Substitute.For<ISignalHub>();
            var logger = Substitute.For<ILoggerFactory>();
            var service = new SignalService<string>(reader, parser, hub, logger);

            //act
            await service.StartAsync(CancellationToken.None);

            //assert
            await hub.Received().SendMessage(Arg.Is<Message>(m => m.Status == "unit test exception"));
        }

        [Fact]
        public async Task StopsReadingOnStop()
        {
            //arrange
            var reader = Substitute.For<ISignalReader<string>>();
            var parser = Substitute.For<ISignalParser<string>>();
            var hub = Substitute.For<ISignalHub>();
            var logger = Substitute.For<ILoggerFactory>();
            var service = new SignalService<string>(reader, parser, hub, logger);

            //act
            var task = service.StartAsync(CancellationToken.None);
            await service.StopAsync(CancellationToken.None);
            await task;

            //assert
            Assert.True(task.IsCompleted);
        }

        [Fact]
        public async Task SetsSignalsOnUpdate()
        {
            //arrange
            var reader = Substitute.For<ISignalReader<string>>();
            var parser = Substitute.For<ISignalParser<string>>();
            var hub = Substitute.For<ISignalHub>();
            var signals = new List<Signal>
            {
                new Signal
                {
                    SSID = "UnitTest",
                    Frequency = Frequency._2_4_GHz,
                    Strength = -30
                }
            };
            parser.Parse(Arg.Any<string>()).Returns(signals);
            var logger = Substitute.For<ILoggerFactory>();
            var service = new SignalService<string>(reader, parser, hub, logger);

            //act
            var task = service.StartAsync(CancellationToken.None);
            await task;

            //assert
            await hub.Received().SendMessage(Arg.Is<Message>(m => m.Signals.Equals(signals)));
        }
    }
}

File: WiFiHeatMap.App.Tests/SignalServiceTests.ts
import { Test, TestSuite } from "xunit.ts";
import SignalService from "../WiFiHeatMap.App/SignalService";
import Signal from "../WiFiHeatMap.App/Signal";
import Message from "../WiFiHeatMap.App/Message";
import { HubConnection } from "@microsoft/signalr";
import Mockito from "ts-mockito";

export default class SignalServiceTests extends TestSuite {
    @Test()
    async signalsAreUpdatedOnNewMessage() {
        //arrange
        const connection = Mockito.mock<HubConnection>();
        Mockito.when(connection.start()).thenResolve();

        const signal1 = new Signal('mac1', 'ssid1', 2, -50);
        const signal2 = new Signal('mac2', 'ssid2', 2, -40);
        const signals = [signal1, signal2];
        const service = new SignalService(Mockito.instance(connection), signals);

        const message = new Message();
        const new1 = new Signal('mac1', 'ssid1', 2, -45);
        const new2 = new Signal('mac2', 'ssid2', 2, -44);
        message.signals = [new1, new2];

        //act
        service.update(message);

        //assert
        this.assert.equal(-45, signals[0].strength);
        this.assert.equal(-44, signals[1].strength);
    }
}